### PR TITLE
Add get_compilation_info

### LINKED ISSUE #2130: Unable to handle shader compilation errors
`wgpu` currently has no API to inspect errors when compiling a shader module. On desktop, `Device::create_shader_module` will panic on error. On web, the different browsers will print out error info to the console, but execution continues, and errors will usually occur later during pipeline creation with the errored `ShaderModule` objects.

The [WebGPU spec](https://www.w3.org/TR/webgpu/#dom-gpushadermodule-compilationinfo) requires a `GPUShaderModule.compilationInfo()` method that should return any errors generated during compilation.

Ideally, either:
 * change `Device::create_shader_module` to idiomatically return a `Result`,
 * or match the spec and add a `ShaderModule::compilation_info` method. The user can then inspect this info and return an error as appropriate.

### PR DESCRIPTION
**Pull request by stefnotch**
**Connections**
Fix #2130

**Description**
This adds an async `get_compilation_info` method to the `wgpu` crate, which either uses the native compilation info as reported by naga, or the compilation info that the browser's WebGPU backend reported.

The general idea is to note down all the relevant info when the compilation happens, and later extract them when the user calls `get_compilation_info`
- WebGPU backend with WGSL code: We store the Javascript promise, and later access it. We also store the source code, so that we can remap the reported offsets.
- WebGPU backend with other code: We translate the code to WGSL, and store the compilation and validation errors. And then we store the Javascript promise, so that we can also report any errors that the underlying WGSL compiler reports. Here, we don't keep the source code around, cause it'd be nonsensical. (Offsets are pointing at auto-generated WGSL code)
- Native backend: We immediately compile it, and store the errors.

Relevant bits about API design are
- Using UTF-8 for positions. We're purposefully deviating from the `GPUCompilationInfo` specification, which uses UTF-16 for reporting positions.
- For any given error, we're only reporting the main error location. If naga reports [multiple locations for a single error](https://github.com/gfx-rs/wgpu/blob/82fa5801525e7a32118ba3e02b9ce4adfb57112a/naga/src/front/wgsl/error.rs#L610), then the additional locations will be in the message.
  - For tooling that really wants to get all the error locations: Please directly use naga

Also moved the `ShaderError` struct from `wgpu_core` to `naga`, so that it can be used by both
- wgpu_core
- webgpu with certain feature flags

Any suggestions on how to simplify the code, or improve it are very appreciated. There are a few bits that feel like I'm duplicating code, or feel suboptimal.

**Testing**

Basic unit tests have been added.

**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `cargo clippy`. If applicable, add:
  - [x] `--target wasm32-unknown-unknown`
  - [ ] `--target wasm32-unknown-emscripten` I cannot get this to run locally without a million compile errors. I clearly missed some setup part.
- [ ] Run `cargo xtask test` to run tests.
- [x] Add change to `CHANGELOG.md`. See simple instructions inside file.

### CODE REVIEW

File: wgpu/src/backend/webgpu.rs
Comment (stefnotch) [inline wgpu/src/backend/webgpu.rs]: This is a Javascript promise. It has been causing me a lot of headaces, because it forces an asynchronous API.

File: wgpu/src/backend/wgpu_core.rs
Comment (stefnotch) [inline wgpu/src/backend/wgpu_core.rs]: Creating a compilation info from a wgpu-core error seems reasonably straightforward, see implementation below.

File: wgpu/src/backend/wgpu_core.rs
Comment (stefnotch) [inline wgpu/src/backend/wgpu_core.rs]: I wasn't entirely sure if I should only take the first span, or do something different.

File: wgpu/src/backend/wgpu_core.rs
Comment (stefnotch) [inline wgpu/src/backend/wgpu_core.rs:3036]: The WebGPU spec says
> A shader-creation error is an error feasibly detectable at shader module creation time. Detection relies only on the WGSL module source text and other information available to the createShaderModule API method. Statements in this specification that describe something the program must do generally produce a shader-creation error if those assertions are violated.

I figured that this is in the spirit of the specification, but am happy to change it.

File: wgpu/src/lib.rs
Comment (stefnotch) [inline wgpu/src/lib.rs:877]: Was creating a new type here the correct move, or should I just have exposed the already existing Naga `Span` type?

Also, the WebGPU spec wants a `lineNum` and a `linePos`. Should I add those to this struct and thus redundantly store them? Or add a helper method to compute them? Or neither of those options?

File: wgpu/src/lib.rs
Comment (stefnotch) [inline wgpu/src/lib.rs]: Piping through the wgpu-core compilation message is easy enough.

But I have no idea how to deal with the Javascript promise and async Rust. Any pointers there would be hugely appreciated.

File: wgpu/src/backend/wgpu_core.rs
Comment (stefnotch) [inline wgpu/src/backend/wgpu_core.rs]: The CI fails, because there's a configuration where need this trait in `wgpu/src/backend/webgpu.rs`, but `#[cfg(wgpu_core)]` doesn't apply. So it doesn't get compiled, and Rust rightfully throws an error.

Any suggestions on where to move it?

File: wgpu/src/lib.rs
Comment (Wumpf) [inline wgpu/src/lib.rs]:
```suggestion
/// except that there is only a `span`, with a byte positions in the UTF-8 encoded source code that was passed for the corresponding shader.
```

File: wgpu-core/src/pipeline.rs
Comment (Wumpf) [inline wgpu-core/src/pipeline.rs]: not entirely clear from the name that this is a source code range this error is about (how about `source_range`?) but more importantly, I'm confused about how this relates to `location` just above. Isn't that what we're actually looking for here?

File: wgpu/src/backend/webgpu.rs
Comment (Wumpf) [inline wgpu/src/backend/webgpu.rs]: nice idea to chain that error simply in :+1:

File: wgpu/src/backend/webgpu.rs
Comment (Wumpf) [inline wgpu/src/backend/webgpu.rs]: since you asked about less code dupl - this one could have a mapping utility function. A `From`/(`Into`) would be nice, but won't work unfortunately since each type is defined elsewhere :/

File: wgpu/src/backend/webgpu.rs
Comment (Wumpf) [inline wgpu/src/backend/webgpu.rs]: that one could be handled with an `From` implementation :)

File: wgpu/src/backend/webgpu.rs
Comment (Wumpf) [inline wgpu/src/backend/webgpu.rs]: err.to_string just concats the strings for all underlying glsl parsing errors if I follow it correctly. So why not do `err.inner.errors.map` and map each error to a `CompilationMessage`?

File: wgpu/src/backend/webgpu.rs
Comment (Wumpf) [inline wgpu/src/backend/webgpu.rs]: as before, why merge the errors and then take arbitrarily the first span instead of exposing all errors als separate messages

EDIT: ah I think this is different here because this is a `WithSpan<ValidationError>` if I'm not mistaken. Hmm still strange though to pick the first span and ignore the rest :/. Maybe nothing to be done about it right now.

File: wgpu/src/backend/webgpu.rs
Comment (Wumpf) [inline wgpu/src/backend/webgpu.rs]: I don't think we should panic if the webgpu impl messes this up. let's return max int and call `log::error!` instead

File: wgpu/src/backend/webgpu.rs
Comment (Wumpf) [inline wgpu/src/backend/webgpu.rs]: let's go with error - after all it's an error in wgpu or the webgpu impl if we don't know about the type :)

**Review by Wumpf — CHANGES_REQUESTED** [review]
Looking great already! Left a few comments on things to improve / check upon.
Very nicely documented api on wgpu :+1:

Regarding tests: Yes plz! :). You can put them under tests/test s, maybe even part of tests/tests/shader, haven't checked if that makes sense. Would be great to have simple test cases for success, warning and error.

Also, please add a changelog entry. This is a great addition that shouldn't be missed there :)

**Comment by stefnotch** [comment]
By the way, if I run
```
cargo check --target wasm32-unknown-emscripten
```
I get the following output. Am I doing something wrong?
```                                                                                                                                                                                                                            04/12/2024 10:45:49 AM AM
    Checking spirv v0.3.0+sdk-1.3.268.0
    Checking bytemuck v1.15.0
    Checking winit v0.29.15
    Checking player v0.19.3 (D:\Coding\GitHub\Libraries\wgpu\player)
error: The platform you're compiling for is not supported by winit
  --> C:\Users\Stefnotch\.cargo\registry\src\index.crates.io-6f17d22bba15001f\winit-0.29.15\src\platform_impl\mod.rs:67:1
   |
67 | compile_error!("The platform you're compiling for is not supported by winit");
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
```

**Comment by Wumpf** [comment]
uh yeah emscripten target seems to be broken again (much worse for me on trunk actually), ignore that for now. `wasm32-unknown-unknown` still seems to work fine (it also gets a lot more testing on ci)

File: wgpu-core/src/pipeline.rs
Comment (stefnotch) [reply wgpu-core/src/pipeline.rs]: I added a separate method, since the existing `location` method requires a source string (and piping that one through is tough), and it computes a few pieces of information that I just end up throwing away. (line number and offset in the line)

If there's a better option that ultimately leads to simpler code, I'd love to hear about it.

File: wgpu-core/src/pipeline.rs
Comment (stefnotch) [inline wgpu-core/src/pipeline.rs]: This one feels odd, most of the error types report one error, right? Meanwhile ParsingGlsl seems to be able to return multiple errors?

File: wgpu/src/backend/webgpu.rs
Comment (stefnotch) [reply wgpu/src/backend/webgpu.rs]: True, good point. It feels a tad inconsistent that the glsl error type has multiple underlying errors while the other error types only seem to have one.
Should anything be done about that?

File: wgpu/src/backend/webgpu.rs
Comment (stefnotch) [reply wgpu/src/backend/webgpu.rs]: Yeah, I'm not certain what the best solution here is. I could
- do nothing and keep it as is
- add a Vec<> for the compilation spans to the `CompilationMessage` type, deviating from the standard
- find a few good examples and open an issue in the repository where the work on the WebGPU specification happens (surely there are cases where an error clearly has multiple spans, right?)
- report multiple spans as multiple errors? But that seems wrong
- join the spans? Also seems wrong

**Comment by stefnotch** [comment]
I added tests for this 🎉  They're very barebones, but should be decent enough to verify that this feature isn't completely broken.

Out of curiosity, do those tests need to be gpu tests? Couldn't they theoretically run with a fake GPU backend?

**Comment by Wumpf** [comment]
> Out of curiosity, do those tests need to be gpu tests? Couldn't they theoretically run with a fake GPU backend?

most of what ci runs is actually with software rasterizers by the respective apis. You're right that for these tests we don't actually need to go to a real backend but we're lacking the necessary mocking architecture to stub out wgpu_hal out of these.

File: wgpu-core/src/pipeline.rs
Comment (Wumpf) [reply wgpu-core/src/pipeline.rs]: I see, makes sense!

File: wgpu-core/src/pipeline.rs
Comment (Wumpf) [reply wgpu-core/src/pipeline.rs]: well the others are also returning iterators and you're only calling `next()` on them, thus only looking at the first one. All of them have several error spans.

cc: @teoxoy generally we should expose these errors one by one to webgpu right?

File: wgpu/src/backend/webgpu.rs
Comment (Wumpf) [reply wgpu/src/backend/webgpu.rs]: as commented above, I don't think glsl is that special here; it just has a slightly different interface in the errors. Also, it's not unlikely that all parsers expose multiple errors in the future.

**Comment by Wumpf** [comment]
huh, sorry for the long delay - I missed the updates on this

Thanks for the improvements, looking better already. But I'm still not happy about any of the "pick first" code in here. If any of these were to remain we need a good (documented) reason each, otherwise I have to assume that we have a bug where we're accidentally hiding errors.

File: wgpu-core/src/pipeline.rs
Comment (teoxoy) [reply wgpu-core/src/pipeline.rs]: Right now only the GLSL frontend reports multiple errors but the WGSL frontend will as well in the future.

WGSL errors are richer and contain multiple spans so that we can point to different parts of the shader, see for example the redefinition error:

https://github.com/gfx-rs/wgpu/blob/82fa5801525e7a32118ba3e02b9ce4adfb57112a/naga/src/front/wgsl/error.rs#L610

Comment (teoxoy) [reply wgpu-core/src/pipeline.rs]: Is this new method necessary? It seems to be doing the same thing as the one above.

Comment (stefnotch) [reply wgpu-core/src/pipeline.rs]: > WGSL errors are richer and contain multiple spans so that we can point to different parts of the shader, see for example the redefinition error:

Thank you for the example! How do I best report that as a WebGPU error?
I'm basically struggling with the fact that the official WebGPU `GPUCompilationMessage` only lets one include a single location
https://developer.mozilla.org/en-US/docs/Web/API/GPUCompilationMessage

One could
- add a `Vec<>` for the compilation spans to the wgpu `CompilationMessage` type, deviating from the standard
- open an issue in the repository where the work on the WebGPU specification happens, and making the case for "multiple locations make sense"
- report multiple spans as multiple errors (e.g. Make the first one an error, and the other ones warnings that pretty much say "error also happened here")



> Is this new method necessary? It seems to be doing the same thing as the one above.

I added a separate method, since the existing location method requires a source string, and it computes a few pieces of information that I just end up throwing away. (line number and offset in the line)

I'd be happy about better suggestions though. Maybe I could tweak the existing `location` methods to return a `Range` and defer resolving the `naga::SourceLocation` until a bit later?

Comment (teoxoy) [reply wgpu-core/src/pipeline.rs]: > How do I best report that as a WebGPU error?
I'm basically struggling with the fact that the official WebGPU `GPUCompilationMessage` only lets one include a single location

I think a single location is fine, as for the error message we can use the `ParseError.emit_to_string()` method (it will return a nicely formatted error message that takes into account all the spans).

The implementation of `.location()` seems to use the first span which I think is correct? for now? I guess it makes sense that the first span would point to the issue at hand and the others are complementary but it doesn't seem to be documented.

> I added a separate method, since the existing location method requires a source string, and it computes a few pieces of information that I just end up throwing away. (line number and offset in the line)

Why do we need to throw some of the fields away? All `naga::SourceLocation`'s fields are part of `GPUCompilationMessage`.

Comment (stefnotch) [reply wgpu-core/src/pipeline.rs]: @teoxoy
Sounds good.

Designing a good `CompilationMessage` for wgpu is a bit annoying, because we pretty much have to deviate from [the standard](https://gpuweb.github.io/gpuweb/#gpucompilationmessage) for it to be idiomatic Rust. Here's what I'm leaning towards at the moment, does this make sense?
```rust
pub struct CompilationMessage {
    pub message: String,
    // `message_type` is a name that doesn't clash with a reserved keyword, unlike `type`
    pub message_type: CompilationMessageType,
    /// An option, instead of the "Must be 0 if the message does not correspond to any specific point in the shader code." rule.
    pub location: Option<SourceLocation>,
}

// Exists so that we can put the location info in an option
pub struct SourceLocation {
    /// Assumes a UTF-8 encoded source instead of assuming UTF-16
    pub offset: u32,
    /// Assumes a UTF-8 encoded source instead of assuming UTF-16
    pub length: u32,
    /// This one actually matches the standard. Well, except that the name is `line_number` instead of `line_num`.
    pub line_number: u32,
    /// Uses Unicode code units, which are more comprehensible for humans instead of UTF-16 offsets.
    /// Naga does the same thing, hence this decision. I'm not sure if it's ideal though. 
    /// e.g. VSCode's column indicator actually uses UTF-16 to count, as far as I can tell. 
    /// And Unicode code units also aren't perfect for humans, since they don't match up with visual characters (graphemes) 
    pub line_position: u32,
}
```

Would love some feedback.

File: wgpu/src/backend/webgpu.rs
Comment (stefnotch) [reply wgpu/src/backend/webgpu.rs]: The other spans should be in the error message, which is hopefully good enough.

File: naga/src/front/glsl/error.rs
Comment (stefnotch) [inline naga/src/front/glsl/error.rs]: Would it make sense to rename this struct below to `ParseErrors`?

File: wgpu-core/src/pipeline.rs
Comment (teoxoy) [reply wgpu-core/src/pipeline.rs]: I think that looks reasonable! I also don't know what would be best for `line_position`; we can go with "unicode code units" for now.

**Comment by kpreid** [comment]
> Using Unicode code units for the line_position. 

>  `/// Uses Unicode code units, which are more comprehensible for humans instead of UTF-16 offsets.`

I think you mean code _points_ here. Counting code units is encoding-dependent. But beyond that terminological nitpick: surely on basic principles this should be counting UTF-8 bytes (aka UTF-8 code units) instead? That's the only way a Rust string can be indexed/sliced efficiently.

If the goal is to mark spans by counting "columns", then Unicode code points do not do that either — you need to execute the [`unicode-width`](https://crates.io/crates/unicode-width) algorithm to approximate that, or ask your font/terminal for feedback to get it perfect. I don't think that should be baked into the API, so it's better to present what is maximally useful for slicing, and let the caller compute the width it cares about.

Don't count code points, because there are almost no use cases where it is the _correct_ answer.

**Comment by stefnotch** [comment]
@kpreid Yes, my bad, I meant code points.

I would, for a brief moment, have argued that counting code points makes sense, since VSCode seems to do that. Except that that's not true, since VSCode has a special case for tabs `\t`. Surprisingly, VSCode doesn't have a special case for emojis that consist of multiple code points. 

So I'm realising that it's bonkers and you absolutely have a point. And that the best possible option is outputting something that a computer can reasonably convert to whatever is most appropriate for a given case. 

Though at that point, one would usually directly use the `offset` field instead?

For what it's worth, I did copy it from the internal naga `SourceLocation` struct. Maybe the internal struct should also be changed?

**Comment by stefnotch** [comment]
Okay, I fixed a few more issues. Importantly, I moved the `ShaderError` struct from `wgpu_core` to `naga`, so that it can be used by both
- wgpu_core
- webgpu with certain feature flags

I wonder if there's a simpler variation of those `cfg`s for conditional compilation than whatever I chose.

File: CHANGELOG.md
Comment (Wumpf) [inline CHANGELOG.md]: now that I already did the 0.20 prep for changelog, could you move this to the highlight section above `New features` with a brief usage example? Deserves one if it should make the cut for 0.20 :)

File: naga/src/front/glsl/error.rs
Comment (Wumpf) [reply naga/src/front/glsl/error.rs]: agreed `ParseErrors` (plural) makes more sense

File: naga/src/front/wgsl/error.rs
Comment (Wumpf) [inline naga/src/front/wgsl/error.rs]: If you remove the todo you practically did it? 😄

File: wgpu/src/backend/webgpu.rs
Comment (Wumpf) [inline wgpu/src/backend/webgpu.rs]: you should check for the `naga` alias here instead, see https://github.com/gfx-rs/wgpu/blob/trunk/wgpu/build.rs#L16

File: wgpu/src/lib.rs
Comment (Wumpf) [inline wgpu/src/lib.rs]: nice, so much cleaner with these `From` implementations all listed out here!

The only thing I findodd is that they work on reference types. I expect from/into to be consuming - if the callsite needs it longer (for error handling as it happens in a few places) it's the callsite's job to do the necessary clone. You're also not entirely consistent with this here having `impl From<naga::SourceLocation> for SourceLocation` further down.

File: wgpu/src/backend/wgpu_core.rs
Comment (Wumpf) [inline wgpu/src/backend/wgpu_core.rs]: might have missed this earlier, but similar to comment on `impl From<...> for SourceLocation` I'd prefer to avoid `From` implementations that operate on references. From/Into is supposed to take ownership of the thing it converts and callsites can get weird & confusing when it sometimes does and sometimes doesn't. (it still consumes the _reference_, but no longer the object itself)

**Review by Wumpf — CHANGES_REQUESTED** [review]
Looking really good now. A few cosmetic notes that would be nice to address.
Thanks @teoxoy for the input - now that this touched Naga stuff more directly I'll wait for your additional sign-off before merge.

Could you try to avoid rebasing on the next iteration? Rebasing may look nice on the history but makes it a lot harder to incrementally review - I can't tell github to show me the changes since last time and instead have to start from scratch

File: naga/src/front/wgsl/error.rs
Comment (stefnotch) [reply naga/src/front/wgsl/error.rs]: Perfect, did it. 

I also checked all the `ParseError`s that get constructed. They all obey the rule, except for the `Error::CyclicDeclaration` case. The cyclic declaration case doesn't really have a "primary span".

File: wgpu/src/lib.rs
Comment (stefnotch) [reply wgpu/src/lib.rs]: I see. I'm guessing that convention is more widespread (like `into_iter` and `iter`) and should be worth remembering.

Will fix :)

**Comment by stefnotch** [comment]
@kpreid @teoxoy I looked into the linePos (column number) a bit more. I also looked into the naga `SourceLocation` struct, since it also contains such info.

The `naga::SourceLocation` struct is mainly being used for emitting SpirV OpLine instructions. Those are for debug tools.
And inconveniently, [the specification](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpLine) says exactly nothing about what the column in an `OpLine` is supposed to represent. It's just supposed to be a positive integer.

So I looked at what some tools do
- SPIRV-Cross: Always emits `0` for the column number.
- DirectXShaderCompiler: [Uses the length of the string plus one](https://github.com/microsoft/DirectXShaderCompiler/blob/00cd8233d5d6755dae1218e5f84bafdad83c7fc5/tools/clang/lib/Basic/SourceManager.cpp#L1118). I couldn't definitively figure out which encoding is being used.
- Renderdoc: Couldn't figure out what it assumes, but it doesn't seem to be counting code points.

So my question: May I change the `naga::SourceLocation.line_position` to count UTF-8 code units instead of general Unicode code points? This only makes a difference when dealing with non-ASCII Unicode characters, so it doesn't actually come up very often.

File: CHANGELOG.md
Comment (Wumpf) [inline CHANGELOG.md]: not all messages are errors, so this is a bit misleading

**Comment by teoxoy** [comment]
> May I change the `naga::SourceLocation.line_position` to count UTF-8 code units instead of general Unicode code points?

Sounds good to me. It makes things more consistent.

File: CHANGELOG.md
Comment (stefnotch) [reply CHANGELOG.md]: Good point, thank you for catching this

**Comment by stefnotch** [comment]
And it's ready for a (hopefully) final review. I'm glad that I got to learn a ton about wgpu internals.

Thank you for all the reviews and helping me get this PR into a decent shape. <3

**Review by Wumpf — APPROVED** [review]
all good from my side 👍 

Thanks again for taking this feature on working through all these iterations! :)

**Review by teoxoy — APPROVED** [review]

**Comment by Wumpf** [comment]
uhoh, this went into 0.20 changelog but didn't get release with it

**Pull request merged**
